src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandlerBase.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    internal abstract class CodeActionsHandlerBase
+    {
+        private readonly ICodeFixService _codeFixService;
+        private readonly ICodeRefactoringService _codeRefactoringService;
+
+        // TODO - this should not be liveshare.
+        protected const string RemoteCommandNamePrefix = "_liveshare.remotecommand";
+        protected const string RunCodeActionCommandName = "Roslyn.RunCodeAction";
+        protected const string ProviderName = "Roslyn";
+
+        public CodeActionsHandlerBase(ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService)
+        {
+            _codeFixService = codeFixService ?? throw new ArgumentNullException(nameof(codeFixService));
+            _codeRefactoringService = codeRefactoringService ?? throw new ArgumentNullException(nameof(codeRefactoringService));
+        }
+
+        public async Task<IEnumerable<CodeAction>> GetCodeActionsAsync(Solution solution, Uri documentUri, LSP.Range selection, CancellationToken cancellationToken)
+        {
+            var document = solution.GetDocumentFromURI(documentUri);
+            if (document == null)
+            {
+                return ImmutableArray<CodeAction>.Empty;
+            }
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
we should open work item to clean up code in codefixservice, coderefactoringservice, suggested actions, suggested source and this.

Reply:
so that we can share more code between them.

Reply:
https://github.com/dotnet/roslyn/projects/45#card-20873061